src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
+    getBroadcastId: GetBroadcastId;
+
+    broadcasts: sb3.Target["broadcasts"];
+    tempo: number;
+    textToSpeechLanguage: TextToSpeechLanguage;
+    videoState: "on" | "off";
+    videoTransparency: number;
+  }
+
+  function serializeStage(stage: Stage, options: SerializeStageOptions): sb3.Stage {
+    // Serialize a stage. Extending from a serialized target, the stage carries
+    // additional properties for values shared across the project - notably,
+    // the broadcast dictionary, as well as values for some extensions.
+
+    const { broadcasts, getBroadcastId } = options;
+    return {
+      ...serializeTarget(stage, { broadcasts, getBroadcastId, stage }),
+      isStage: true,
+      tempo: options.tempo,
+      textToSpeechLanguage: options.textToSpeechLanguage,
+      videoState: options.videoState,
+      videoTransparency: options.videoTransparency
+    };
+  }
+
+  function serializeProject(project: Project): sb3.ProjectJSON {
+    // Serialize a project. This is the master function used when project.toSb3
+    // is called. The main purpose of serializeProject is to serialize each
+    // target (sprite or stage) and collect them together in the final output
+    // format. It also provides utility functions shared across every target's
+    // serialization, e.g. broadcast utilities.
+
+    // Maintain broadcast dictionaries in both name <-> id directions. sb-edit
+    // operates on broadcast names, but Scratch 3.0 largely requires access to
+    // their IDs as well, so we generate IDs for names as they are required.
+    // Name -> ID is for converting sb-edit names and is the backbone behind
+    // getBroadcastId; ID -> name is the format which is directly serialized
+    // on the stage object. (Only the stage need carries a filled-out broadcast
+    // dictionary in Scratch 3.0.)
+    let broadcastNameToId: { [name: string]: string } = {};
+    let broadcastIdToName: { [id: string]: string } = {};
+
+    const getBroadcastId = (name: string): string => {
+      if (!(name in broadcastNameToId)) {
+        const id = generateId();
+        broadcastNameToId[name] = id;
+        broadcastIdToName[id] = name;
+      }
+      return broadcastNameToId[name];
     };
-    return sb3Sprite;
-  });
+
+    // Preemptively parse through all the broadcast inputs in the project and
+    // generate a broadcast ID for each of them.
+    for (const target of [project.stage, ...project.sprites]) {
+      for (const block of target.blocks) {
+        if (block.opcode === OpCode.event_whenbroadcastreceived) {
+          if (block.inputs.BROADCAST_OPTION.type === "broadcast") {
+            getBroadcastId(block.inputs.BROADCAST_OPTION.value);
+          }
+        } else if (block.opcode === OpCode.event_broadcast || block.opcode === OpCode.event_broadcastandwait) {
+          if (block.inputs.BROADCAST_INPUT.type === "broadcast") {
+            getBroadcastId(block.inputs.BROADCAST_INPUT.value);
+          }
+        }
```

Comment:
```suggestion
        if (
          block.opcode === OpCode.event_whenbroadcastreceived ||
          block.opcode === OpCode.event_broadcast ||
          block.opcode === OpCode.event_broadcastandwait
        ) {
          const broadcastInput = (block.opcode === OpCode.event_whenbroadcastreceived) ?
            block.inputs.BROADCAST_OPTION :
            block.inputs.BROADCAST_INPUT;

          if (broadcastInput.type === "broadcast") {
            getBroadcastId(broadcastInput.value);
          }
        }
```

This looks more DRY to me but it's a matter of preference.

Reply:
Also, under what conditions is the `type` not `"broadcast"`?